### PR TITLE
Implement minimum account age requirement and temp blocks

### PR DESCRIPTION
**Pull request by Taaku18**
Addresses #175 #120

# Added
- Introduced a new configuration variable `account_age` for setting a minimum account creation age.
  - Users blocked by this reason will be stored in `blocked` along with other reasons for being blocked.
  - `account_age` needs to be an ISO8601 Duration Format (examples: `P12DT3H` 12 days and 3 hours, `P3Y5M` 3 years and 5 months `PT4H14M999S` 4 hours 14 minutes and 999 seconds). https://en.wikipedia.org/wiki/ISO_8601#Durations.

# Changed
- `block` reason cannot start with `System Message: ` as it is now reserved for internal user blocking.
- `block`, like `close`, now support a block duration (temp blocking).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,14 +4,21 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-# v2.13.7
+# [Unreleased]
 
 ### Added
+- Addresses #175.
```
Comment (fourjr) [inline CHANGELOG.md]: this line isn't necessary but if you want to put it, hyperlink to original issue

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,14 +4,21 @@ All notable changes to this project will be documented in this file.
...
+- Introduced a new configuration variable `account_age` for setting a minimum account creation age.
+  - This utilizes `blocked` configuration variable to keep track of users under the limit.
```
Comment (fourjr) [inline CHANGELOG.md]: technical details aren't necessary

File: bot.py [hunk 3]
```diff
@@ -437,27 +437,67 @@ def main_color(self):
...
+            account_age = float(account_age)
+            if account_age < 0 or not account_age.is_integer():
+                raise ValueError
+        except ValueError:
+            logger.warning('The age limit needs to be a whole number, '
+                           f'not "{account_age}".')
+            del self.config.cache['account_age']
+            await self.config.update()
+        else:
+            account_age = int(account_age)
```
Comment (fourjr) [inline bot.py]: why not leave it as a float?

File: bot.py [hunk 4]
```diff
@@ -437,27 +437,67 @@ def main_color(self):
...
-                pass
+                logger.warning(info(f'Blocked emoji ({blocked_emoji}) '
+                                    'is not a valid emoji.'))
+                del self.config.cache['blocked_emoji']
+                await self.config.update()
 
-        if str(message.author.id) in self.blocked_users:
+        account_age = self.config.get('account_age', 0)
+        try:
+            account_age = float(account_age)
```
Comment (fourjr) [inline bot.py]: would there be an error if account_age is a str?

File: CHANGELOG.md (hunk 1 above)
Comment (Taaku18) [reply CHANGELOG.md]: Removed.

File: bot.py (hunk 3 above)
Comment (Taaku18) [reply bot.py]: It's left as a float now.

File: bot.py (hunk 4 above)
Comment (Taaku18) [reply bot.py]: No because it's caught by the try-except block.

Comment (fourjr) [reply bot.py]: my bad

File: CHANGELOG.md (hunk 2 above)
Comment (Taaku18) [reply CHANGELOG.md]: Modified that message to make it less technical.

File: bot.py (hunk 3 above)
Comment (Taaku18) [reply bot.py]: Now uses the international time standard ISO8601 duration format for `account_age`.

File: bot.py [hunk 5]
```diff
@@ -431,33 +433,102 @@ def main_color(self):
...
+        return sent_emoji, blocked_emoji
 
-        if sent_emoji not in UNICODE_EMOJI:
+    async def process_modmail(self, message):
+        """Processes messages sent to the bot."""
+        sent_emoji, blocked_emoji = await self.retrieve_emoji()
+
+        account_age = self.config.get('account_age')
+        if account_age is None:
+            account_age = isodate.duration.Duration()
```
Comment (fourjr) [inline bot.py]: `account_age = config.get() or isodate.duration.Duration()`

File: bot.py [hunk 6]
```diff
@@ -431,33 +433,102 @@ def main_color(self):
...
+            await self.config.update()
+            min_account_age = message.author.created_at
+
+        if min_account_age > datetime.utcnow():
+            # user account has not reached the required time
             reaction = blocked_emoji
+            changed = False
+
+            if str(message.author.id) not in self.blocked_users:
+                new_reason = 'System Message: New Account.'
```
Comment (fourjr) [inline bot.py]: Reason should be clearer. Suggestion: "System Message: User account age too low, below required {required_age}".

Then use a `startswith` later to check

File: cogs/modmail.py [hunk 7]
```diff
@@ -706,7 +740,12 @@ def format_log_embeds(self, logs, avatar_url):
     @trigger_typing
     @checks.has_permissions(manage_channels=True)
     async def unblock(self, ctx, *, user: User = None):
-        """Unblocks a user from using Modmail."""
+        """
+        Unblocks a user from using Modmail.
+
+        Note: reasons start with "System Message: " are reserved for internal
```
Comment (fourjr) [inline cogs/modmail.py]: `reasons that start with`

File: cogs/modmail.py [hunk 8]
```diff
@@ -718,17 +757,30 @@ def format_log_embeds(self, logs, avatar_url):
...
             del self.bot.config.blocked[str(user.id)]
             await self.bot.config.update()
-            embed = discord.Embed(
-                title='Success',
-                color=self.bot.main_color,
-                description=f'{mention} is no longer blocked'
-            )
+
+            if msg.startswith('System Message: '):
+                reason = msg[16:].strip('. \n\r')
```
Comment (fourjr) [inline cogs/modmail.py]: Add a (python) comment here explaining the strip 🤔

File: bot.py (hunk 5 above)
Comment (kyb3r) [reply bot.py]: `config.get('account_age', isodate.duration.Duration())`

Comment (Taaku18) [reply bot.py]: I did that for the else clause directly follows that

File: bot.py (hunk 6 above)
Comment (Taaku18) [reply bot.py]: Now it's: `System Message: New Account. Required to wait for 1 year, 7 months and 2 days.`

File: cogs/modmail.py (hunk 7 above)
Comment (Taaku18) [reply cogs/modmail.py]: The sentence in its current state is grammatically correct. The "that" is just personal preference. I personally like to exclude "that" when possible, but I guess I'll add it anyways.

File: cogs/modmail.py (hunk 8 above)
Comment (Taaku18) [reply cogs/modmail.py]: Meant to remove trailing spaces and periods, no need for comment, changed it to `msg.strip().rstrip('.')`.

File: CHANGELOG.md [hunk 9]
```diff
@@ -18,17 +30,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 # v2.13.8
 
 ### Fixed
+- A bug where a thread was blocked from sending messages when multiple images were uploaded, due to a typo.
 
-- A bug where a thread was blocked from sending messsages when multiple images were uploaded. (This was due to a typo)
+### changed
```
Comment (fourjr) [inline CHANGELOG.md]: Capital `c` please

**Comment by fourjr** [comment]
Merge this ASAP and start working in #186. That should be the highest priority

**Comment by Taaku18** [comment]
> Merge this ASAP and start working on #186. That should be the highest priority

Sure, merge at your discretion.

**Pull request merged**